fat loot token and daily dungeon

taltamir opened the pull request:
reworked fat loot token acquisition. Primarily daily dungeon:
- removed useMalware() function. it was an outdated duplicate of our daily dungeon code plus setting a combat directive to use maleware. this should have been done via combat function being set to use malware, which actually someone did add some time ago making useMalware() function entirely redundant
- migrate choice handling for daily dungeon from old method (changing the choice adventure setting) to new method which uses auto_choice_adv.ash. in this case it points at void dailyDungeonChoiceHandler(int choice, string[int] options)
- if you do not have a lockpick in the locked door daily dungeon NC use your primestat instead of hardcoding to always use muscle
- in softcore if you do not have or cannot use a cubeling then pull daily dungeon tools which saves about 30 to 40 adv
    - when pulling lockpicks prefer to pull Platinum Yendorian Express Card if available instead of lockpicks
- created and used float estimateDailyDungeonAdvNeeded() instead of just checking if you have 15 adventures. if you have the tools it takes fewer adventures
- boolean LX_dailyDungeonToken() split off from LX_phatLootToken(). the latter has improved logic on when to do daily dungeon and when to do fantasy realm. fantasyrealm is always 5 adv. daily dungeon is 15 adv without the tools and ~3 adv with the tools. So we only want to prioritize daily dungeon over fantasy realm if we have the tools.
- do not assume success when adventuring in daily dungeon. return the result of autoAdv()
-  make sure we have the means to handle choice adventure 692 [I Wanna Be a Door] before adventuring in daily dungeon
- auto_forcePhatLootToken setting added. it lets you force grabbing phat loot tokens every day. mainly used for new accounts who need to farm enough tokens to grab the cubeling and the 3 skills

## How Has This Been Tested?

played with for 2 days on 5 accounts:
3 doing HC Ed. 
1 doing SC lowkey summer accordion thief
1 doing HC oxy seal clubber

## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have based by pull request against the [beta branch](https://github.com/Loathing-Associates-Scripting-Society/autoscend/tree/beta) or have a good reason not to.

taltamir commented:
I am certain there was 1 more thing to do in this PR. but i keep forgetting what it was

Malibu-Stacey commented:
>boolean LX_dailyDungeonToken() split off from LX_phatLootToken(). the latter has improved logic on when to do daily dungeon and when to do fantasy realm

LoL you read my mind. I was thinking about this exact thing the other day for #463

Malibu-Stacey commented on RELEASE/scripts/autoscend/quests/level_any.ash:
```diff
@@ -363,104 +363,156 @@ boolean LX_lockPicking()
 }
 
-boolean LX_phatLootToken()
+float estimateDailyDungeonAdvNeeded()
```
:+1:
There is some way to calculate exactly which encounters you'll receive and the order you'll receive them in since the dungeon is procedurally generated every day (http://noblesse-oblige.org/calendar/ lists the monster encounters every day, I have never seen it be wrong) but this is an excellent alternative.
It has annoyed me when I see my test runs get the cubeling drops late on day 1 in an LKS run and then it doesn't bother doing the dungeon because it has less than 15 adventures constantly (but then I remember LKS runs take 4 days in Normal anyway and I have the reward permed so it doesn't matter).

taltamir replied:
oh nice. I was not aware of this calculator.
thank you.

Malibu-Stacey commented on RELEASE/scripts/autoscend/quests/level_any.ash:
```diff
@@ -363,104 +363,156 @@ boolean LX_lockPicking()
+	
+	// make sure we have enough adventures. since partial completion means wasted adventures.
+	int adv_budget = my_adventures() - auto_advToReserve();
+	if(adv_budget < estimateDailyDungeonAdvNeeded())
```
```suggestion
	if(adv_budget < ceil(estimateDailyDungeonAdvNeeded()))

```

taltamir commented on RELEASE/scripts/autoscend/auto_choice_adv.ash:
```diff
@@ -136,6 +137,55 @@ boolean auto_run_choice(int choice, string page)
+			{
+				run_choice(1);	// take damage to progress
+			}
+			break;
```
change made

taltamir commented on RELEASE/scripts/autoscend/quests/level_any.ash:
```diff
@@ -363,104 +363,156 @@ boolean LX_lockPicking()
 	{
-		backupSetting("choiceAdventure690", 3);
-		backupSetting("choiceAdventure691", 3);
+		providePlusNonCombat(25, true);
```
change made

taltamir commented:
@Malibu-Stacey I applied all the changes you asked.
also I added:
* when checking if we have enough adventures, check adv < 1+estimate just to be safe in case the estimate is on the low end. This is probably not necessary
* added a new setting auto_forcePhatLootToken that lets you force grabbing the token every day even if you have enough. helps new accounts get the cubeling faster

Malibu-Stacey commented on BUILD/settings/any.dat:
```diff
@@ -48,3 +48,4 @@ auto_secondPlaceOrBust	boolean	When true, abort before each tower test if we can
 auto_helpMeMafiaIsSuperBrokenAaah	boolean	Pretty much just for werebear. If for some reason you find that mafia completely fails to detect item drops occasionally, leading to wasting a ton of adventures continuing to look for an item you already found, try this out.
 auto_logLevel	string	One of: critical, error, warning, info, debug, trace. Sets the level of log output for autoscend gcli and session log output. Critical showing the least detail (only critical messages) and trace showing the most detail (print everything). Defaults to info.
 auto_restoreUseBloodBond	boolean	Whether to use extra hp to cast blood bond. Blood bond has a recurring HP drain, set to false if you are worried about getting killed or wasting resources healing. Defaults to false.
+auto_forcePhatLootToken	boolean	force grabbing the phat loot tokens from daily dungeon and fantasy realm every day even if you already have enough for this run. This is mainly for new accounts who want to get the cubeling.
```
```suggestion
auto_forcePhatLootToken	boolean	force grabbing the phat loot tokens from daily dungeon and fantasy realm every day even if you already have enough for this run. This is mainly for new accounts who want to get the cubeling and/or skillbooks.
```

Malibu-Stacey commented on RELEASE/scripts/autoscend/autoscend_header.ash:
```diff
@@ -23,7 +23,10 @@ boolean LM_edTheUndying();						//Defined in autoscend/auto_edTheUndying.ash
+float estimateDailyDungeonAdvNeeded();				//Defined in autoscend/quests/level_any.ash
+boolean LX_phatLootToken();							//Defined in autoscend/quests/level_any.ash
+boolean LX_dailyDungeonToken();						//Defined in autoscend/quests/level_any.ash
+void dailyDungeonChoiceHandler(int choice, string[int] options);	//Defined in autoscend/quests/level_any.ash
```
could you move this to the end of the file with the other choice handlers please?

taltamir replied:
change made

Malibu-Stacey commented on RELEASE/scripts/autoscend/quests/level_any.ash:
```diff
@@ -363,104 +363,215 @@ boolean LX_lockPicking()
+				run_choice(1);	// take damage to progress
+			}
+			break;
+		default:
+			break;
```
```suggestion
		default:
			abort("unhandled choice in dailyDungeonChoiceHandler");
			break;

```

Malibu-Stacey commented:
I think you need to run the build.sh to update the autoscend_settings.txt file too btw.

taltamir commented:
@Malibu-Stacey all requested changes made
